Schema API

rorbech opened the pull request:
Adding public schema API. 

```
interface BaseRealm : Versioned {
    fun schema(): RealmSchema
}

interface RealmSchema {
    val classes: Collection<RealmClass>
    operator fun get(key: String): RealmClass?
}

interface RealmClass { // Matches realm_class_info_t except that attributes are also attached here
    val name: String
    val properties: Collection<RealmProperty>
    val primaryKey: RealmProperty? // equivalent to:  properties.any { it.primaryKey }
    operator fun get(key: String): RealmProperty?
}

interface RealmProperty { // Matches realm_property_info_t
    val name: String
    val type: RealmPropertyType
    // Referring to actual Kotlin KType property nullability and not Core's nullability flag
    val isNullable: Boolean
}

sealed interface RealmPropertyType {
    val storageType: RealmStorageType
    val isNullable: Boolean
}

class ValuePropertyType(
    override val storageType: RealmStorageType,
    override val isNullable: Boolean,
    val isPrimaryKey: Boolean,
    val isIndexed: Boolean
) : RealmPropertyType

class ListPropertyType(
    override val storageType: RealmStorageType,
    override val isNullable: Boolean = false
) : RealmPropertyType
```

Closes #85 
Closes #238 
Closes #201 (Added final test of embedded indicator in schema retrieved from Core)

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmSchema.kt:
```diff
@@ -0,0 +1,86 @@
+    val name: String // val publicName: String or should we just support mappings already?
+    val index: Boolean
+    val primaryKey: Boolean
+    val type: RealmPropertyType
```
Not sure this is enough to also cover collection types? I think we talked about using the same bitfield approach that Core uses, but maybe there are other ways of doing it 🤔

rorbech commented:
@nhachicha @cmelchior I have tried to update according to discussion, but there are some annoying things around initializing the schema details from the compiler plugin. 

I wrapped it as an implementation detail of `RealmClass` in https://github.com/realm/realm-kotlin/pull/539/files#diff-af15d86f1d0cfbfd742b652cd227e5a386558f19cf2b968714d4915812e9dffdR27. But it still feels annoying to initialize the details according to the C-interops instead of using the abstraction of flags, etc. as in https://github.com/realm/realm-kotlin/pull/539/files#diff-66ad0c5e3b9031ba0b834a0d7e8c081238e8ce69ff0ccc5a2dd0a1d9a14a4ed0R26

Besides of the current code not working on Darwin (🤦 ... I am looking into that...),  my gut feeling would be that we would be better off using the clean abstraction of RealmClassImpl with the penalty of having to initialize the `Table/Property` on open. At least until we know that it is an issue or we get another enlightenment from caching the keys/updating schema on `realm_schema_changed`-callbacks. 

It would maybe even be safer if this was going through a factory method in the SDK, so that it would be more flexible to update if we were changing conventions. Has there been any considerations/issues with that in realm-java?

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/BaseRealmImpl.kt:
```diff
@@ -62,6 +64,14 @@ abstract class BaseRealmImpl internal constructor(
         log.info("Realm opened: ${configuration.path}")
     }
 
+    // FIXME Currently constructs a new instance on each invocation. We could cache this pr. schema
```
I would probably prefer caching it. It should also be possible inside the Notifier thread.

rorbech replied:
Caching it requires that it is update along with the `RealmReference` to work on Kotlin native. I think implementing this goes better in hand with implementation of key caching. Created #567 to track that.

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmClass.kt:
```diff
@@ -0,0 +1,26 @@
+
+interface RealmClass { // Matches realm_class_info_t except that attributes are also attached here
+    val name: String
+    val properties: Collection<RealmProperty>
```
Any thinking behind using Collection here? would maybe expect List or Set to be more appropriate?

rorbech replied:
Sets doesn't offer random access apis while lists doesn't signal uniqueness of elements. But reiterating it I think that I removed the random access apis, so I guess we could just go with sets again. 

I don't know if there are different clear expectations between sets/list that would be nice to keep in mind (constant order, etc.) and if it is only me being annoyed about using List if we guarantee uniqueness of elements?

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmSchema.kt:
```diff
@@ -0,0 +1,22 @@
+package io.realm.schema
+
+interface RealmSchema {
+    val classes: Collection<RealmClass>
```
Why Collection here? Instead of Set or List?

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmSchema.kt:
```diff
@@ -0,0 +1,22 @@
+
+interface RealmSchema {
+    val classes: Collection<RealmClass>
+    operator fun get(key: String): RealmClass
```
Shouldn't this be `RealmClass?` ?

rorbech commented on packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/RealmInterop.kt:
```diff
@@ -74,6 +77,10 @@ expect object RealmInterop {
+    fun realm_get_class_keys(realm: NativePointer): List<ClassKey>
+    fun realm_find_class(realm: NativePointer, name: String): ClassKey
+    fun realm_get_class(realm: NativePointer, classKey: ClassKey): Table
+    fun realm_get_class_properties(realm: NativePointer, classKey: ClassKey, max: Long): List<Property>
```
The C-API requires allocation of the out-array of keys up front, and the count is often already known from retrieving the class info else where, so just trying to save a roundtrip. Should we rather just embed a call to fetch the class info inside?

rorbech commented on packages/cinterop/src/darwin/kotlin/io/realm/internal/interop/RealmEnums.kt:
```diff
@@ -53,21 +53,32 @@ actual enum class PropertyType(override val nativeValue: UInt) : NativeEnumerate
-    RLM_PROPERTY_PRIMARY_KEY(realm_wrapper.RLM_PROPERTY_PRIMARY_KEY),
-    RLM_PROPERTY_INDEXED(realm_wrapper.RLM_PROPERTY_INDEXED),
+actual object PropertyFlags {
+    actual val RLM_PROPERTY_NORMAL: Int = realm_wrapper.RLM_PROPERTY_NORMAL.toInt()
```
This is purely internal and is just a one-to-one mapping of core values. I guess _none_ is correct, but itches to change a single of the entries and already tried to hide the actual flags behind an abstraction with https://github.com/realm/realm-kotlin/pull/539/files#diff-bc9f8abf28b6e500356cbcd4f9682782d925c78eab42949716570d9a54681598R37-R39 and https://github.com/realm/realm-kotlin/pull/539/files#diff-bc9f8abf28b6e500356cbcd4f9682782d925c78eab42949716570d9a54681598R50-R52

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/ElementType.kt:
```diff
@@ -0,0 +1,33 @@
+    // This corresponds to realm_property_type in realm.h but naming would overlap with
+    // RealmPropertyType
+    // Mimics storage type. This is important to user if we do custom type maps, etc.
+    enum class FieldType {
```
If this is too convoluted we can also try to go with the below and hope that it is clear that the `nullable` is relating to the elements in the collection. After all the confusion in realm-java mostly arises from the `@Required` annotation being on the member, while it for Kotlin is obvious that the nullability is on the elements with `?`-indicator in the type parameter ... ?
```
data class RealmPropertyType(
    val collectionType: RealmCollectionType,
    val dataType: RealmDataType,
    val nullable: Boolean
)```

rorbech replied:
Tried to rework the type information according to discussions. Think it came our quite nice except #naming ... open for suggestions.

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/CollectionType.kt:
```diff
@@ -0,0 +1,27 @@
+
+package io.realm.schema
+
+// Should we try to avoid enums completely as introducing new ones breaks compatibility due to
```
Yeah. Think we can just go with the enums. Maybe interesting to follow https://youtrack.jetbrains.com/issue/KT-38750 to be able to selectively required/disable exhaustiveness for specific enums.

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/ElementType.kt:
```diff
@@ -0,0 +1,33 @@
+
+package io.realm.schema
+
+// FIXME Probably dangerous to have data classes in public API
```
Mostly a note to my self that I should re-iterate this. We could have interfaces and just do internal implementations by data classes. But, I guess it makes sense to have as a public data class for this specific case, as people might want to construct instances to test against the schema they retrieve from the API.

rorbech commented on packages/cinterop/src/darwin/kotlin/io/realm/internal/interop/RealmInterop.kt:
```diff
@@ -380,6 +385,90 @@ actual object RealmInterop {
+                outCount.ptr
+            )
+            if (outCount.value.toLong() < 1) {
+                error("Invalid schema: Class without properties")
```
No. Just returning empty lists instead.

rorbech commented on packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/Table.kt:
```diff
@@ -16,12 +16,14 @@
-// FIXME API-SCHEMA Platform independent class definition. Maybe rework into utility method called in Realm
-//  object's companion schema mechanism depending on how we relate this to the actual schema/runtime
-//  realm_class_info_t.
+// FIXME NAMING I think the consistent naming should be the OS names, Class/Property, but guess we
```
I tried to clean up naming using `ClassInfo` and `PropertyInfo` for the cinterop classes that is also basically a 1-1 mapping to the C-API structs.

rorbech commented:
I tried to rework the property type information after our discussions. Please have another look.

cmelchior commented on test/base/src/androidTest/kotlin/io/realm/test/shared/RealmSchemaTests.kt:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    @Test
+    fun realmClass() {
```
Missing test where the class isn't part of the schema.

rorbech replied:
Added `realmClass_notFound` and `realmProperty_notFound`

cmelchior commented on test/base/src/androidTest/kotlin/io/realm/test/shared/RealmSchemaTests.kt:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    @Test
+    fun realmProperty() {
```
Missing test where property wasn't found

cmelchior commented on test/base/src/commonMain/kotlin/io/realm/entities/schema/SchemaVariations.kt:
```diff
@@ -0,0 +1,45 @@
+import io.realm.entities.Sample
+import io.realm.realmListOf
+
+class SchemaVariations : RealmObject {
```
Maybe add a comment about what this class does? Because while it does cover schema "options", it doesn't cover all permutations.

cmelchior commented on test/base/src/androidTest/kotlin/io/realm/test/shared/RealmSchemaTests.kt:
```diff
@@ -0,0 +1,183 @@
+        assertEquals(2, schema.classes.size)
+
+        // Verify properties of SchemaVariations
+        val classDescriptor = schema["SchemaVariations"] ?: fail("Couldn't find class")
```
Maybe use `SchemaVariations::class.simpleName`?

cmelchior commented on test/base/src/androidTest/kotlin/io/realm/test/shared/RealmSchemaTests.kt:
```diff
@@ -0,0 +1,183 @@
+
+    @Test
+    @Suppress("NestedBlockDepth")
+    fun schema_optionCoverage() {
```
Maybe add a comment about this test is for ensuring that we have a class that touches all possible options for a property but _doesn't_ cover all permutations... I was a bit confused about this at least.

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmSchema.kt:
```diff
@@ -0,0 +1,22 @@
+package io.realm.schema
+
+interface RealmSchema {
+    val classes: Collection<RealmClass>
```
Should we also expose the schema version here? Or does that causes problems during migrations 🤔

rorbech replied:
I also iterated the thought a couple of times, but didn't find the magic bullet. I think we should expose it on an immutable schema, but since I haven't landed the _schema refresh_ parts of the key-based accessors, I am in doubt on the exact requirements and what would make sense. 

Currently the compiler plugin builds the embedded schema metadata directly with the cinterop-types and haven't change to use the public API as container during compile time, but if we change to use the public API types through out compiler plugin too, then we would have to be able to represent a schema without a _transaction version_ (input is welcome). 

Will add a todo and see if I can settle it as part of #600

cmelchior replied:
Yes, we can always come back to it once other parts settle. Just adding a TODO is 👍

cmelchior reviewed (changes requested):
Nice ...the refactor was definitely for the better IMO.

I have some minor touchpoints:
- Missing Dokka
- Some naming
- A little bit about tests

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmStorageType.kt:
```diff
@@ -0,0 +1,26 @@
+    STRING,
+    OBJECT,
+    FLOAT,
+    DOUBLE;
```
I also just merged TIMESTAMP support, sorry 😄

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/BaseRealm.kt:
```diff
@@ -23,6 +25,13 @@ interface BaseRealm : Versioned {
     val configuration: RealmConfiguration
 
+    /**
+     * Returns a copy of the schema of the realm at invocation time.
```
It was just to make it explicit that it reflects the schema at invocation time, since `Realm.schema()` can yield different schemas over time. Maybe 
```
Returns an immutable schema of the realm.
```
A bit less convoluted, but still indicates that the instance doesn't change. ?

cmelchior replied:
Yes, that wording is bettter IMO.

rorbech commented on packages/jni-swig-stub/realm.i:
```diff
@@ -27,7 +27,24 @@ using namespace realm::jni_util;
+                java.lang.reflect.Method loadMethod = classToLoad.getDeclaredMethod("load");
+                loadMethod.invoke(instance);
+            } catch (Exception e) {
+                e.printStackTrace();
```
It is java code, so need to catch the checked exceptions. Will rethrow as a `RuntimeException`

cmelchior replied:
Ah yeah, true. The reflection API's throw checked exceptions...but yeah, rethrowing would be better.

rorbech commented on packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/PropertyInfo.kt:
```diff
@@ -0,0 +1,69 @@
+            isIndexed: Boolean
+        ): PropertyInfo {
+            val flags =
+                (if (isNullable) RLM_PROPERTY_NULLABLE else 0) or (if (isPrimaryKey) RLM_PROPERTY_PRIMARY_KEY else 0) or (if (isIndexed) RLM_PROPERTY_INDEXED else 0)
```
What is it that you are missing? https://github.com/realm/realm-core/blob/master/src/realm.h#L306

cmelchior replied:
Ah, I thought these flags was these ones: https://github.com/realm/realm-core/blob/master/src/realm/object-store/property.hpp#L46... if not, just ignore my comment 👍

rorbech replied:
Yeah, all these flags are split out into _property type_, _collection type_ and various flags in the C-API

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/schema/RealmPropertyImpl.kt:
```diff
@@ -0,0 +1,57 @@
+    override var type: RealmPropertyType,
+) : RealmProperty {
+
+    override val isNullable: Boolean = when (type) {
```
Yes

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmPropertyType.kt:
```diff
@@ -0,0 +1,46 @@
+    val isIndexed: Boolean
+) : RealmPropertyType
+
+class ListPropertyType(
```
Yes, that was the thought.

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmStorageType.kt:
```diff
@@ -0,0 +1,26 @@
+
+package io.realm.schema
+
+enum class RealmStorageType {
```
🎉 Life time achievement. I succeeded in naming something ... but maybe you incepted it 😞

rorbech replied:
I also think that the term `RealmStorageType` will make good sense in combination with type maps.

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmPropertyType.kt:
```diff
@@ -0,0 +1,46 @@
+        //  allows to define the options centrally and use it to verify exhaustiveness in tests.
+        //  JUST DON'T FORGET TO UPDATE ON WHEN ADDING NEW SUBCLASSES :see_no_evil:
+        //  We could do a JVM test that verifies that it is exhaustive :thinking:
+        val subTypes: Set<KClass<out RealmPropertyType>> = setOf(SingularPropertyType::class, ListPropertyType::class)
```
Yep. More of a building block for building the metadata as with the `TypeDescriptor` members. I will internalize for now

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmPropertyType.kt:
```diff
@@ -0,0 +1,46 @@
+    }
+}
+
+class SingularPropertyType(
```
I think `Singular` is the correct meaningful term, but I guess it doesn't sound that good as part of a type name. I tried with `ValuePropertyType` 🤷

rorbech commented:
@cmelchior Updated naming, added Dokka (though I am not sure I got the vocabulary right on all the meta description, schema, realm object, realm object model, etc. 🤯) and adjusted and commented the test a bit. 

Regarding the exhaustiveness of tests: I think it makes sense to just focus on the codepaths of the schema API and leave the exhaustiveness tests to the various modeling feature tests. Besides that, I think it will require too much restructure of migrating the various `TypeDescriptor` members into the various schema API classes and would probably fit better in a separate PR. I get that this exercise might uncover some details/feedback to the API, but the API seems relatively simple so I envision that it is more additive than breaking.

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmPropertyType.kt:
```diff
@@ -0,0 +1,65 @@
+/**
+ * A [ValuePropertyType] describes singular value properties.
+ */
+data class ValuePropertyType(
```
I like this name, but after talking to @edualonso , we might have small design headache around queries on objects and primitive lists that will require us to have two implementations. In that case, we should probably align whatever we end up calling the query classes with this.

edualonso replied:
Sigh... 😢

cmelchior reviewed (approved):
Nice 💯 ... only had a few minor comments.

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmProperty.kt:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Returns whether or not the property is allowed to be null in the corresponding `RealmObject`
+     *
+     * For non-[ValuePropertyType] this is different from [RealmPropertyType.isNullable]  which
```
The special case is the `RealmProperty.isNullable` and not the `RealmProperty.type.isNullable`, so think it would be confusing to place it in the `ListPropertyType`. 

```
For [ValuePropertyType] this will always be the same as [RealmPropertyType.isNullable] and 
for all other types it will be false.
```

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmClass.kt:
```diff
@@ -0,0 +1,49 @@
+     * @returnn the class' primary key [RealmProperty] or `null` it the class doesn't have a primary
+     * key.
+     */
+    // If this is never to change we could make it a val, but if using same API for mutable schema
```
Since we don't have the outline of the mutable schema I ended up exposing it as a property.

clementetb commented on packages/cinterop/src/jvm/kotlin/io/realm/internal/interop/PropertyType.kt:
```diff
@@ -29,9 +29,9 @@ actual enum class PropertyType(override val nativeValue: Int) : NativeEnumerated
-            return values().find { it.nativeValue == i } ?: error("Unknown type: $i")
+    actual companion object {
+        actual fun from(nativeValue: Int): PropertyType {
+            return values().find { it.nativeValue == nativeValue } ?: error("Unknown type: $nativeValue")
```
```suggestion
            return values().find { it.nativeValue == nativeValue } ?: error("Unknown property type: $nativeValue")
```

clementetb commented on packages/cinterop/src/jvm/kotlin/io/realm/internal/interop/CollectionType.kt:
```diff
@@ -23,5 +23,11 @@ actual enum class CollectionType(override val nativeValue: Int) : NativeEnumerat
+
+    actual companion object {
+        actual fun from(nativeValue: Int): CollectionType {
+            return values().find { it.nativeValue == nativeValue } ?: error("Unknown type: $nativeValue")
```
```suggestion
            return values().find { it.nativeValue == nativeValue } ?: error("Unknown collection type: $nativeValue")
```

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/schema/RealmStorageTypeImpl.kt:
```diff
@@ -0,0 +1,34 @@
+            io.realm.internal.interop.PropertyType.RLM_PROPERTY_TYPE_FLOAT -> RealmStorageType.FLOAT
+            io.realm.internal.interop.PropertyType.RLM_PROPERTY_TYPE_DOUBLE -> RealmStorageType.DOUBLE
+            io.realm.internal.interop.PropertyType.RLM_PROPERTY_TYPE_TIMESTAMP -> RealmStorageType.TIMESTAMP
+            else -> error("Unknown type: $type")
```
```suggestion
            else -> error("Unknown storage type: $type")
```

clementetb reviewed (approved):
LGTM, great work!!! Added some minor comments.

rorbech commented on packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/ClassInfo.kt:
```diff
@@ -16,16 +16,18 @@
-    val flags: Set<PropertyFlag> = setOf(PropertyFlag.RLM_PROPERTY_NORMAL)
-)
+    val primaryKey: String?,
+    val numProperties: Long,
```
I agree that count sounds better, but since this is a 1-1 representation of the similar Core struct `realm_class_info` I think it is ok to keep the names aligned.

rorbech commented on packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/ClassInfo.kt:
```diff
@@ -16,16 +16,18 @@
-)
+    val primaryKey: String?,
+    val numProperties: Long,
+    val numComputerProperties: Long = 0,
```
Same. Will just keep the names aligned with the Core struct

clementetb reviewed (approved):
LGTM
